platform/sha/generic/mbedtls/src/mender-sha.c

```diff
@@ -0,0 +1,63 @@
+/**
+ * @file      mender-sha.c
+ * @brief     Mender SHA interface for MbedTLS platform
+ *
+ * Copyright Northern.tech AS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mender-sha.h"
+#include "mender-log.h"
+
+#include <mbedtls/sha256.h>
+
+mender_err_t
+mender_sha256_begin(mender_sha256_context_t *context) {
+    assert(NULL != context);
+
+    mbedtls_sha256_context *ctx = malloc(sizeof(mbedtls_sha256_context));
+    if (NULL == ctx) {
+        mender_log_error("Unable to allocate memory");
+        return MENDER_FAIL;
+    }
+
+    mbedtls_sha256_init(ctx);
+    mbedtls_sha256_starts(ctx, 0); /* Use SHA-256, not SHA-224 */
+
+    *context = ctx;
+    return MENDER_OK;
+}
+
+mender_err_t
+mender_sha256_update(mender_sha256_context_t context, const unsigned char *input, size_t length) {
+    assert(NULL != context);
+
+    mbedtls_sha256_context *ctx = context;
+    mbedtls_sha256_update(ctx, input, length);
+
+    return MENDER_OK;
+}
+
+mender_err_t
+mender_sha256_finish(mender_sha256_context_t context, unsigned char *output) {
+    mbedtls_sha256_context *ctx = context;
+    if (NULL != ctx) {
+        if (NULL != output) {
```

Comment:
See doc string in the declaration ;)

Reply:
Yes, but why would we allow this use case?

Reply:
So that if there occurs an error, you can still release the context without having to write a checksum to a buffer. Furthermore, it's common convention that functions releasing resources accept NULL as the argument.